### PR TITLE
feat: Create Polynomial payout curve

### PR DESCRIPTION
**Pull request by dillionverma**
### What?

Add Polynomial payout curve to node-dlc.

### Why?

To support custom strategy oracles which have linear payouts.

### CODE REVIEW

**Comment by matthewjablack** [comment]
Would be good to update README as well regarding DLC Spec Compliance now that we have polynomial payout curve support: https://github.com/AtomicFinance/node-dlc#dlc-spec-compliance

**Comment by dillionverma** [comment]
> Would be good to update README as well regarding DLC Spec Compliance now that we have polynomial payout curve support: https://github.com/AtomicFinance/node-dlc#dlc-spec-compliance

good point, just updated!

File: packages/core/lib/dlc/PolynomialPayoutCurve.ts
Comment (matthewjablack) [inline packages/core/lib/dlc/PolynomialPayoutCurve.ts:138]: Shouldn't `payoutFunction.pieces.length` only be `2` if we're dealing with a line? 

And it would be 3 for Quadratic, and 4 for Cubic

Line: https://github.com/bitcoin-s/bitcoin-s/blob/master/core/src/main/scala/org/bitcoins/core/protocol/dlc/models/DLCPayoutCurve.scala#L421

**Review by matthewjablack — COMMENTED** [review]
Couple questions for this PR ☝️

File: packages/core/lib/dlc/PolynomialPayoutCurve.ts
Comment (zpv) [reply packages/core/lib/dlc/PolynomialPayoutCurve.ts:138]: @matthewjablack 
image

The three pieces are for the max loss, actual payout curve, and max gain (3 lines in total)

Comment (zpv) [reply packages/core/lib/dlc/PolynomialPayoutCurve.ts:138]: A **line curve piece** consists of two points.
There are three line curve pieces that make up our payout function.

Comment (dillionverma) [reply packages/core/lib/dlc/PolynomialPayoutCurve.ts:138]: Great question! Yeah as Steven mentioned, our `PayoutFunction` consists of three separate `PolynomialPayoutCurve`'s (in order to set a max loss and max gain), and each curve is made up of two points.

File: packages/core/lib/utils/Precision.ts
Comment (dillionverma) [reply packages/core/lib/utils/Precision.ts:11]: Great question, had to think a bit again about this one. The `getPrecision` and `toPrecision` are methods based on what is defined in the specs [here](https://github.com/discreetlogcontracts/dlcspecs/blob/master/PayoutCurve.md#version-0-payout_function)

As defined in the specs

> Precision is set to be the first 16 bits of the payout after the binary point which were rounded away

So in the case of `BigNumber(1000.1)`, the `precision` value is `1000000000000000` since the only digit after the decimal point is 1 and the number itself can hold up to 16 digits.

Another example is `getPrecision(new BigNumber(1.112233445566778899))` which returns `1122334455667788` which is the first 16 digits after the decimal point.

File: packages/core/lib/dlc/PolynomialPayoutCurve.ts
Comment (matthewjablack) [reply packages/core/lib/dlc/PolynomialPayoutCurve.ts:138]: Ah I see it's representing lines not points. Makes sense. Thanks for clarification.

File: packages/core/lib/utils/Precision.ts
Comment (matthewjablack) [reply packages/core/lib/utils/Precision.ts:11]: Ah that makes more sense

Here they are referring to `extraPrecision` meaning digits that come after the decimal. Not precision in terms of significant digits. Would be good to add a comment here to clarify.

**Review by matthewjablack — APPROVED** [review]
Would be good to add those comments regarding extra precision for clarification

Also would be good to get review from @zpv since he's more familiar with this part of the codebase

Other than that LGTM 🎉

File: packages/core/lib/dlc/PolynomialPayoutCurve.ts
Comment (zpv) [inline packages/core/lib/dlc/PolynomialPayoutCurve.ts:137]: What about in the case where there's a single line from (0, max_outcome)?
You would only require one curve piece.

Other cases:
- max loss defined, but not max gain    _/
- max gain, but no max loss   /‾‾

Ideally, this function should accept any number (>0) of curve pieces.
For our case, check if `splitIntoRanges` supports a flat line and simply outputs one payout cet payout:

```
[{
         payout: payoutFunction.pieces[0].endpointPayout,
         indexFrom: payoutFunction.endpoint0,
         indexTo: payoutFunction.pieces[0].endpoint,
       }]
```

if so, we can just map `splitIntoRanges` over the provided curve pieces

**Comment by zpv** [comment]
The PR looks good for our use case.

Just one nice-to-have as a suggestion: where we should technically be able to support arbitrary number of curve pieces in `computePayouts`

If `splitIntoRanges` doesn't play nicely w/ a flat line, we can just support only our use case for now :)

**Pull request merged**
